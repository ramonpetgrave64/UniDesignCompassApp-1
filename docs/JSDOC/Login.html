<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <title>JSDoc: Class: Login</title>

    <script src="scripts/prettify/prettify.js"> </script>
    <script src="scripts/prettify/lang-css.js"> </script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <link type="text/css" rel="stylesheet" href="styles/prettify-tomorrow.css">
    <link type="text/css" rel="stylesheet" href="styles/jsdoc-default.css">
</head>

<body>

<div id="main">

    <h1 class="page-title">Class: Login</h1>

    




<section>

<header>
    
        <h2><span class="attribs"><span class="type-signature"></span></span>Login<span class="signature">()</span><span class="type-signature"></span></h2>
        
            <div class="class-description">Component that handles User login and displays Login GUI</div>
        
    
</header>

<article>
    <div class="container-overview">
    
        

    
    <h2>Constructor</h2>
    

    
    <h4 class="name" id="Login"><span class="type-signature"></span>new Login<span class="signature">()</span><span class="type-signature"></span></h4>
    

    















<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="pages_Login.js.html">pages/Login.js</a>, <a href="pages_Login.js.html#line25">line 25</a>
    </li></ul></dd>
    

    

    

    
</dl>




















    
    </div>

    

    

    

    

    

    

    
        <h3 class="subsection-title">Members</h3>

        
            
<h4 class="name" id="createAccount"><span class="type-signature"></span>createAccount<span class="type-signature"></span></h4>




<div class="description">
    Navigates to the Signup page when the create account button is clicked
</div>







<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="pages_Login.js.html">pages/Login.js</a>, <a href="pages_Login.js.html#line77">line 77</a>
    </li></ul></dd>
    

    

    

    
</dl>






        
            
<h4 class="name" id="handleAuth"><span class="type-signature"></span>handleAuth<span class="type-signature"></span></h4>




<div class="description">
    Handles user authentication by sending user information to cognito and graphql database.
</div>







<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="pages_Login.js.html">pages/Login.js</a>, <a href="pages_Login.js.html#line43">line 43</a>
    </li></ul></dd>
    

    

    

    
</dl>






        
            
<h4 class="name" id="handleChange"><span class="type-signature"></span>handleChange<span class="type-signature"></span></h4>




<div class="description">
    Changes state information dependent on the input information from front end
</div>







<dl class="details">

    

    

    

    

    

    

    

    

    

    

    

    

    
    <dt class="tag-source">Source:</dt>
    <dd class="tag-source"><ul class="dummy"><li>
        <a href="pages_Login.js.html">pages/Login.js</a>, <a href="pages_Login.js.html#line37">line 37</a>
    </li></ul></dd>
    

    

    

    
</dl>






        
    

    

    

    
</article>

</section>




</div>

<nav>
    <h2><a href="index.html">Home</a></h2><h3>Classes</h3><ul><li><a href="ChooseStructure.html">ChooseStructure</a></li><li><a href="Compass.html">Compass</a></li><li><a href="CreatePage.html">CreatePage</a></li><li><a href="Dashboard.html">Dashboard</a></li><li><a href="DescribePhase.html">DescribePhase</a></li><li><a href="Graph.html">Graph</a></li><li><a href="LogCard.html">LogCard</a></li><li><a href="Login.html">Login</a></li><li><a href="NewPassword.html">NewPassword</a></li><li><a href="Phase.html">Phase</a></li><li><a href="PhaseNumber.html">PhaseNumber</a></li><li><a href="Process.html">Process</a></li><li><a href="SignUp.html">SignUp</a></li><li><a href="SubmitCompass.html">SubmitCompass</a></li><li><a href="Verification.html">Verification</a></li></ul><h3>Global</h3><ul><li><a href="global.html#LogHandler">LogHandler</a></li></ul>
</nav>

<br class="clear">

<footer>
    Documentation generated by <a href="https://github.com/jsdoc/jsdoc">JSDoc 3.6.2</a> on Thu May 23 2019 18:37:54 GMT-0400 (Eastern Daylight Time)
</footer>

<script> prettyPrint(); </script>
<script src="scripts/linenumber.js"> </script>
</body>
</html>