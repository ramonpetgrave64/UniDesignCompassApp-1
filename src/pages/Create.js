/**
 * @fileoverview The Create Page that allows logged in users to create their Compasses
 * @author Abraham Villaroman
 * @version 1.0.0
*/

import React from "react";
import { navigate } from 'gatsby';
import Layout from '../components/layout';
import { Button, Form } from 'react-bootstrap';
import _ from "lodash";
import {createNewCompass} from '../graphql_utils/utils';
import { connect } from 'react-redux';
import { updateUser } from '../state/actions'
import SpinningWheel from "../components/SpinningWheel"

/**
 * @class SubmitCompass
 * Form that handles submission of the design process
 * takes in the title of the desired compass and fires the createCompass handler
 */
export const SubmitCompass = (props) =>  (
  <Form.Group controlId="formBasicEmail">
    <Form.Label className="form-title">Title of your Design Compass</Form.Label>
    <div className="input-area">
      <Form.Control name="title" type="text" placeholder="Title" onChange={props.onChange} value={props.title} />
      <Button id="create-button" className="input-button" variant="primary" onClick={props.createCompass}> Submit Compass </Button>
    </div>
  </Form.Group>
)

/**
 * @class Choose Structure
 * Form that handles Structure of the design process
 * prompts the user to either pick the default structure or a custom structure
 */
export const ChooseStructure = (props) =>  (
    <Form.Group controlId="formBasicEmail">
      <Form.Label className="form-title">Choose a Compass Style:</Form.Label>
      <div className="input-area">
        <Button className="input-button" id="Custom"  variant="primary" onClick={e => {props.handleCompassType(false)}}> Custom </Button>
        <Button className="input-button" id="Default" variant="primary" onClick={e => {props.handleCompassType(true)}}> Default </Button>
      </div>
      
    </Form.Group>
)
const default_phases= [
  {title:"Define a Problem",
   description:"Phase for defining the problem"},
   {title:"Research",
   description:"researching our problem"},
   {title:"Brainstorm",
   description:"brain storming solutions"},
   {title:"Select",
   description:"selecting the right solutions"},
   {title:"Construct",
   description:" constructing the solution"},
   {title:"Evaluate",
   description:" evaluating the implmentation of the solution"},
   {title:"Communicate",
   description:"test solution with others"},
   {title:"Redesign",
   description:"redesigning thesolution"}
]

/**
 * @class PhaseNumber
 * Form that handles Structure of the design process
 * prompts the user to choose the number phases they want in their process
 */
export const PhaseNumber = (props) =>  (
  <Form.Group controlId="exampleForm.ControlSelect1">
    <Form.Label className="form-title">How many phases are in your design process</Form.Label>
    <div className="input-area">
      <Form.Control as="select" onChange={props.handlePhaseAmount}>
        <option value="0">-</option>
        <option value="2">2</option>
        <option value="3">3</option>
        <option value="4">4</option>
        <option value="5">5</option>
        <option value="6">6</option>
        <option value="7">7</option>
        <option value="8">8</option>
        <option value="9">9</option>
        <option value="10">10</option>
        <option value="11">11</option>
        <option value="12">12</option>
      </Form.Control>
    </div>
  </Form.Group>
)

/**
 * Form that handles creation of a phase in their design process
 * prompts the user to name their phase and it's description
 */
export class DescribePhase extends React.Component{
  state = {
    title: "",
    description: "",
  }

  onChange = (e) => {
    this.setState({
      [e.target.name] : e.target.value 
    },() => {
      this.props.createPhase(this.state.title, this.state.description,this.props.id + 1)
    })

  }

  render(){
    const {title,description} = this.state;
    return (
      <Form.Group >
        <Form.Label className="form-sub-title">{`Phase ${this.props.id + 1}`}</Form.Label>
        <Form.Control id="phaseTitle" name="title" type="text" placeholder="Title" onChange={this.onChange} value={title} />
        <Form.Control id="phaseDescription" name="description" as="textarea" placeholder="Description" onChange={this.onChange} value={description} rows="3" />
      </Form.Group>
    )
  }
} 

/**
 * The page that handles creation of a a design process
 * holds all the logic in changing and storing the form data, then submits the data to the server
 */
export class CreatePage extends React.Component {
  state = {
    compassTitle: "",
    canSubmit: false,
    defaultCompass: true,
    numberOfPhases: 0,
    phases: [],
    status: "chooseStructure",
    loading: false
  }

/**
 * event handler for the Process title
 * changes the value of this.state.compassTitle
 * @param {Object} event object that holds the value of the Process Title
*/
  handleCompassTitle = (e) => {
    this.setState({compassTitle: e.target.value})
  }

/**
 * event handler for the Process Structure
 * changes the value of this.state.defaultCompass
 * when the user submits the structure, switch to the submitCompass form
 * @param {bool} boolean that holds the structure type of our Process (true === default structure, false === custom structure)
*/
  handleCompassType = (isDefaultCompass) => {
    if (isDefaultCompass){
      this.setState({defaultCompass: isDefaultCompass, phases: default_phases,status: "submitCompass"})
    } else {
      this.setState({defaultCompass: isDefaultCompass, status: "numOfPhases"})
    }
  }

/**
 * event handler for the Phase amount 
 * changes the value of this.state.phaseAmount
 * when the user submits the number of phases, switch to the createPhases form
 * @param {Object} event object that holds the value of the Phase Amount
*/
  handlePhaseAmount = (e) => {
    this.setState({numberOfPhases: e.target.value,status:"createPhases"})
  }

/**
 * event handler that creates the phases
 * changes the value of this.state.phases with an array of phases with their title and descriptions
 * @param {Object} event object that holds the Phase objects
*/
  createPhase = (title,description,phaseNumber) => {
    let phases = this.state.phases
    phases.push({title,description,phaseNumber});
    phases = _.uniqBy(phases.reverse(), 'phaseNumber')
    this.setState({phases})
  }
  
/**
 * makes the api request to submit a new compass with: Process Title and the defined phases in the Process
 * if the server request is succesful, the user is redirected to the process screen
 * otherwise, there was an error
*/
  createCompass = () => {
    const {compassTitle, phases} = this.state;
    this.setState({ loading: true }, () => {
      createNewCompass(this.props.user,compassTitle,phases)
      .then((res) => {
          navigate(`/Process/${res.data.getProcess.id}`)
      })
      .catch((err) => {
        alert(err)
        this.forceUpdate();
      })
    })
  }
  /**
   * when the user is done with all the phases, change the form to the submitCompass Form
   */
  submitPhases = (e) => {
    this.setState({status: "submitCompass"})
  }

  /**
   * handles the displaying of forms for the user to operate with to create their design process
   */
  handleForms = () => {
    switch(this.state.status) { 
      case "chooseStructure":
        return <ChooseStructure handleCompassType={this.handleCompassType}/>   
      case "numOfPhases":
        return <PhaseNumber handlePhaseAmount={this.handlePhaseAmount}/>
      case "createPhases":
        const Phases = Array.from({length: this.state.numberOfPhases}, (_, id) => ( 
            <DescribePhase createPhase={this.createPhase} id={id} key={id}/> 
          )
        )
        return (
          <div>
            <Form.Label className="form-title">Describe each phase in your design process</Form.Label>
              {Phases}
            <Button onClick={this.submitPhases}>Submit Phases</Button>
          </div>
        )
      case "submitCompass":
        return <SubmitCompass 
                  createCompass={this.createCompass} 
                  title={this.state.compassTitle} 
                  onChange={this.handleCompassTitle}
                />
      default: 
        return <ChooseStructure handleCompassType={this.handleCompassType}/>   
    }
  }

  render(){
    return (
      <Layout>
          <h2> Create Your Compass </h2>
          <div className="form-container">
            {this.handleForms()}
            { this.state.loading && <SpinningWheel/>}
          </div>
      </Layout>
    )
  }
}
const mapStateToProps =({state})=>(
  {user: state.user}
)
const mapDispatchToProps = dispatch => ({
  updateUser: (user) => dispatch(updateUser(user))
})
export default connect(mapStateToProps,mapDispatchToProps)(CreatePage);
