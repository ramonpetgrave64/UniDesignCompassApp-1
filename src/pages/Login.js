/**
 * @fileoverview The login file responsible for display the login page and handeling login functionality
 * @author <a href="">Ronuel Diaz</a>
 * @version 1.0.0
 */
import React, { Component } from "react";
import { navigate } from "gatsby"
import Layout from "../components/layout"
import "../components/bootstrap.css"
import { connect } from 'react-redux';
import { authenticateUser } from '../state/actions';
import { Auth } from "aws-amplify";
import config from "../aws-exports";
import { getUserbyUsername } from "../graphql_utils/utils";
import SpinningWheel from "../components/SpinningWheel";
import Button from "react-bootstrap/Button";
import FormControl from "react-bootstrap/FormControl";
import FormGroup from "react-bootstrap/FormGroup";
import ControlLabel from "react-bootstrap/FormLabel";
Auth.configure(config);

/**
 * Component that handles User login and displays Login GUI
 */
export class Login extends Component {

    state = {
        password: "",
        username: "",
        user: [],
        loading: false
    };

    /**
     * Changes state information dependent on the input information from front end
     */
    handleChange = (e) => {
        this.setState({ [e.target.id]: e.target.value });
    }
    /**
     * Handles user authentication by sending user information to cognito and graphql database.
     */
    handleAuth = (e) => {
        e.preventDefault();
        this.setState({ loading: true }, () => {
            Auth.signIn(this.state.username, this.state.password)
                .then((res) => {
                    console.log(res)
                    switch (res.challengeName) {
                        case "NEW_PASSWORD_REQUIRED":
                            navigate("/newPassword", { state: { username: this.state.username } });
                            break;
                        case "SOFTWARE_TOKEN_MFA":
                            break;
                        case "MFA_SETUP":
                            break;
                        case "SMS_MFA":
                            break;
                        default:
                            getUserbyUsername(this.state.username)
                                .then((res) => {
                                    const user = res.data.getUser
                                    this.props.authenticateUser(user);
                                    navigate("/")
                                })
                    }
                }, (error) => {
                    console.log(error);
                    alert(error.message);
                    this.setState({ loading: false })
                });
        })
    };
    /**
     * Navigates to the Signup page when the create account button is clicked
     */
    createAccount = (e) => {
        navigate("/SignUp");
    }
    render(){
        return (
            <Layout>
                <div className='container'>
                    <div className="Login">
                        <form onSubmit={this.handleAuth}>
                            <FormGroup controlId="username" bssize="large">
                                <ControlLabel>Username</ControlLabel>
                                <FormControl
                                    placeholder="Username"
                                    autoFocus
                                    type="text"
                                    onChange={this.handleChange}
                                />
                            </FormGroup>
                            <FormGroup controlId="password" bssize="large">
                                <ControlLabel>Password</ControlLabel>
                                <FormControl
                                    placeholder="Password"
                                    onChange={this.handleChange}
                                    type="password"
                                />
                            </FormGroup>
                            <Button 
                                bssize="large"
                                type="submit"
                            >
                                Login
                            </Button>
                            <Button className="float-right" onClick={this.createAccount}>Create New Account</Button>
                        </form>
                    </div>
                </div>
                {this.state.loading && <SpinningWheel />}</Layout>
        )
    }
}

const mapStateToProps = ({ state }) => ({ isAuthenticated: state.isAuthenticated })

const mapDispatchToProps = dispatch => ({
    authenticateUser: (auth) => dispatch(authenticateUser(auth))
})
export default connect(mapStateToProps, mapDispatchToProps)(Login);
