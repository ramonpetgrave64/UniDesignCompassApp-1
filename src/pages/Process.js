/**
 * @fileoverview The Process page that shows the compass and the anayltics of a process
 * @author Abraham Villaroman
 * @version 1.0.0
*/
import React from "react";
import { Tab, Nav } from 'react-bootstrap';
import Layout from "../components/layout";
import {getProcess} from "../graphql_utils/utils" 
import Graph from "../components/Graph";
import Compass from "../components/Compass"
import "../components/bootstrap.css"

/**
 * The Component that handles the display/tabbing between Compass and AnalyticsPage
 * holds all the logic in retrieving the process according to proces id stored in the url
*/
export class Process extends React.Component {

  state = {        
    date_end: "",
    date_start: "",
    id: "",
    name: "",
    phases: [],
    updateComponent: ""
  }

  componentDidMount() {
    const id = this.props.location.pathname.replace(process.env.PROCESS_LINK,"").replace("/","")
    let temp_id = id;
    if(temp_id.indexOf("Process/")!=-1){
    let index_p = temp_id.indexOf("Process/")+8;
    temp_id = temp_id.slice(index_p);
    }
    this.getProcessItems(temp_id);
  }

  /**
   * @param {string} id of the process
   * makes an api request to retrieve all of the data of the process from the process ID
   * once the data is retrieved, change the state of the component
  */
  getProcessItems = (id) => {
    getProcess(id)
    .then((res) => {
      console.log(res);
      const {date_end, date_start, id, name, phaseids : { items }} = res.data.getProcess
      this.update_process_state(date_end, date_start, id, name, items);
    })
  }

  update_process_state = (date_end, date_start, id, name, phaseids)=>{
    this.setState({
      date_end, 
      date_start, 
      id, 
      name, 
      phases: phaseids,
      updateCount: 0
    })
  }

  /**
   * @param {string} either "Compass" or "Graph", the components to refresh
   *  change the state of the component to refresh, when the person chooses a tab, the child component will make a new request to the api for the latest data
  */
  updateHandler = (updateComponent) => {
    this.setState({updateComponent})
  }

  render() {
    const { name,id } = this.state;
    return (
      <Layout>
        <h2 className="text-center">{name}</h2>
        <Tab.Container id="left-tabs-example" defaultActiveKey="compass">
          <Nav variant="pills" defaultActiveKey="compass" className="process-tabs">
            <Nav.Item className="tab">
              <Nav.Link eventKey="compass" onClick={()=>{ this.updateHandler("Compass")}}>Compass</Nav.Link>
            </Nav.Item>
            <Nav.Item className="tab">
              <Nav.Link eventKey="analytics" onClick={()=>{ this.updateHandler("Graph")}}>Analytics</Nav.Link>
            </Nav.Item>
          </Nav>
          <Tab.Content>
            <Tab.Pane eventKey="compass">
              {id && <Compass id={id} updateHandler={this.updateHandler} updateComponent={this.state.updateComponent === "Compass"}/>}
            </Tab.Pane>
            <Tab.Pane eventKey="analytics">
              {id && <Graph processId={id} updateHandler={this.updateHandler} updateComponent={this.state.updateComponent === "Graph"}/>}
            </Tab.Pane>
          </Tab.Content>
        </Tab.Container>
      </Layout>
    );
  }
}

export default Process;
