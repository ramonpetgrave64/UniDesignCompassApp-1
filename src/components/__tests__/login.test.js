import React from "react"
import { Login } from "../../pages/Login";
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { shallow } from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import { ConsoleLogger } from "@aws-amplify/core";
configure({ adapter: new Adapter() });

describe("Login page tests", () => {
    const tree = shallow(<Login />)
    let state
    beforeEach(() => {

        state = {
            password: "",
            username: "",
            user: [],
            loading: false
        };

    });
    describe("Render Tests",()=>{
        it("Snapshot Test",()=>{
            expect(shallowToJson(tree)).toMatchSnapshot();
        });
        it("Check Form controls",()=>{
            expect(tree.find('FormControl').length).toEqual(2);

        })
        it("Check Buttons",()=>{
            expect(tree.find('Button').length).toEqual(2);
        })
    })
    describe("state tests", () => {
        it("Check Page with Blank State", () => {
            expect(tree.state()).toEqual(state)
        })
        
        it("Check input update with non blank State", () => {
            tree.find('FormControl').first().simulate('change',{target:{id:'username',value:'rdiaz002'}});
            tree.find('FormControl').last().simulate('change',{target:{id:'password',value:'rdiaz002'}});
            expect(tree.state().username).toEqual('rdiaz002');
            expect(tree.state().password).toEqual('rdiaz002');
        })
        it("Check loading wheel",()=>{
            tree.setState({loading:false});
            expect(tree.instance())
        })
    })
})

