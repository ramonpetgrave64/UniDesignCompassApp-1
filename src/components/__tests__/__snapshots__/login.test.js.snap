// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login page tests Render Tests Snapshot Test 1`] = `
<Layout>
  <div
    className="container"
  >
    <div
      className="Login"
    >
      <form
        onSubmit={[Function]}
      >
        <FormGroup
          as="div"
          bssize="large"
          controlId="username"
        >
          <FormLabel
            column={false}
            srOnly={false}
          >
            Username
          </FormLabel>
          <FormControl
            as="input"
            autoFocus={true}
            onChange={[Function]}
            placeholder="Username"
            type="text"
          />
        </FormGroup>
        <FormGroup
          as="div"
          bssize="large"
          controlId="password"
        >
          <FormLabel
            column={false}
            srOnly={false}
          >
            Password
          </FormLabel>
          <FormControl
            as="input"
            onChange={[Function]}
            placeholder="Password"
            type="password"
          />
        </FormGroup>
        <Button
          active={false}
          bssize="large"
          disabled={false}
          type="submit"
          variant="primary"
        >
          Login
        </Button>
        <Button
          active={false}
          className="float-right"
          disabled={false}
          type="button"
          variant="primary"
        >
          Create New Account
        </Button>
      </form>
    </div>
  </div>
</Layout>
`;
